Adds a routine to get a LoopKernel reproducer

kaushikcfd opened the pull request:
/cc @jdsteve2

inducer commented on loopy/tools.py:
```diff
@@ -601,4 +610,124 @@ def is_interned(s):
+    python_code = remove_lines_with_only_spaces(
+            remove_common_indentation(python_code))
+
+    # {{{ checking whether the written string is accurate
+
+    reproducer_variables = {}
+    exec(python_code, reproducer_variables)
+    if reproducer_variables["kernel"] != kernel:
+        # knl1 = kernel
+        # knl2 = reproducer_variables["kernel"]
+        raise LoopyError(f"Error generating reproducer for '{kernel.name}'.")
+
+    # }}}
```
Package into a function and use assert?

kaushikcfd replied:
Moved to a function, kept it as a LoopyError 69fc8a8.

inducer commented on loopy/tools.py:
```diff
@@ -601,4 +610,124 @@ def is_interned(s):
 
-# vim: foldmethod=marker
+
+def get_python_reproducer(kernel):
```
Add to docs?

kaushikcfd replied:
Yep, done.

inducer commented on loopy/tools.py:
```diff
@@ -601,4 +610,124 @@ def is_interned(s):
+    """
+    Returns a :class:`str` of a python code that instantiates *kernel*.
+
+    :arg kernel: An instance of :class:`loopy.LoopKernel`
```
Add a
```
.. note::
```
explaining limitations and mention that PRs are welcome.

inducer commented on loopy/tools.py:
```diff
@@ -601,4 +610,124 @@ def is_interned(s):
 
-# vim: foldmethod=marker
+
+def get_python_reproducer(kernel):
```
Rename: `kernel_to_python`?

kaushikcfd replied:
Done 5c71793.

inducer reviewed (commented):
Thx! Some thoughts below.

kaushikcfd commented:
Hi James!
Thanks for trying it out.

Added an argument for the variable name in the reproducer script.
> If information about the imports could also be returned, or the imports could be returned separately as a list (?), then I could more easily check to make sure I'm not adding redundant imports. 

Are you talking about the potentially redundant imports like `import numpy as np` that we emit at the top? I just added it to keep the implementation of this routine simple. We could be smarter about it, but I'm not quite convinced if it's worth making the logic here more convoluted.

inducer commented on loopy/tools.py:
```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
+def _verify_generated_reproducer_is_equivalent(python_code, kernel, var_name):
+    reproducer_variables = {}
+    exec(python_code, reproducer_variables)
+    assert reproducer_variables[var_name] == kernel
```
Move the `assert` out (so that in the `python -O` case, we don't pay the cost of the `exec`?)

kaushikcfd replied:
Done, thanks!

inducer commented on loopy/tools.py:
```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
+    <%! import loopy as lp %>
+    import loopy as lp
+    import numpy as np
+    from pymbolic.primitives import *
```
```suggestion
    import pymbolic.primitives as p
```
Feasible?

kaushikcfd replied:
We emit the shape as: `shape=(Variable('n'), Variable('n'))` as we borrow pymbolic expression's `repr`. And that's why I took this route.

inducer commented on loopy/tools.py:
```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
+
+        options[insn.id] = option[:-2]  # get rid of the trailing ", "
+
+    python_code = r"""
+    <%! import loopy as lp %>
+    import loopy as lp
+    import numpy as np
+    from pymbolic.primitives import *
```
Maybe generate the imports as a separate string (or list of strings)?

kaushikcfd replied:
The routine `t_unit_to_python` now takes in an argument `return_preamble_and_body_separately` which does this. Thanks!

inducer commented on loopy/tools.py:
```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
+        % if isinstance(insn, lp.Assignment):
+        ${insn.assignee} = ${insn.expression} {${opts}}
+        % elif isinstance(insn, lp.BarrierInstruction):
+        ... ${insn.synchronization_kind[0]}barrier{${opts}}
```
```suggestion
        ... ${insn.synchronization_kind[0]}barrier {${opts}}
```

kaushikcfd replied:
Done, thanks!

inducer commented on loopy/tools.py:
```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
+                ),
+            % endfor
+            ],
+            lang_version=${lp.MOST_RECENT_LANGUAGE_VERSION},
```
Why not reproduce the original language version?

kaushikcfd replied:
Loopy kernels do not store language versions. The only way to reproduce lang version would be by copying the entire `LoopKernel.options`, I think that wouldn't be quite good looking.

inducer commented on loopy/tools.py:
```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
+    % endfor
+    """
+
+    python_code = Template(python_code).render(options=options,
```
```suggestion
    python_code = Template(python_code, strict_undefined=True).render(options=options,
```

kaushikcfd replied:
Done, thanks!

inducer reviewed (commented):
A few super minor things below. Otherwise, I'm in favor of just merging this and then improving it as we go (particularly given that it comes with "insurance").

inducer commented on test/test_loopy.py:
```diff
@@ -3074,6 +3074,21 @@ def test_deps_from_conditionals():
+            "{[i]: 0<=i<n}",
+            """
+            y[i] = sin(x[i]) {id=insn}
+            z[i] = sin(y[i]) {dep=insn}
```
Maybe add a reduction here? :)

kaushikcfd replied:
Done, the test now includes that too!

inducer commented:
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

inducer commented:
LGTM. Thanks for working on this!
